test/functional/ex_cmds/read_spec.lua

```diff
@@ -0,0 +1,93 @@
+local t = require('test.testutil')
+local n = require('test.functional.testnvim')()
+
+local eq, write_file, clear, pcall_err = t.eq, t.write_file, n.clear, t.pcall_err
+local fn = n.fn
+local setline, getline, setcharpos, execute = fn.setline, fn.getline, fn.setcharpos, fn.execute
+
+local tmp_file = 'text.txt'
+local original_text = { 'First', 'Last' }
+local read_text = { ' This is a line starts with a space', '  This one starts with two spaces.' }
+local inserted_middle = { original_text[1], read_text[1], read_text[2], original_text[2] }
+local inserted_start = { read_text[1], read_text[2], original_text[1], original_text[2] }
+
+local function test_read(cmd, expected)
+  setline(1, original_text)
+  setcharpos('.', { 0, 0, 0 })
+  execute(cmd)
+  for i, e in ipairs(expected) do
+    eq(e, getline(i))
+  end
+end
+
+local function test_undo(cmd)
+  setline(1, original_text)
+  execute(cmd)
+  execute('undo')
+  eq(original_text, { getline(1), getline(2) })
+end
+
+describe(':read', function()
+  local function cleanup()
+    os.remove(tmp_file)
+  end
+  before_each(function()
+    clear()
+    cleanup()
+    write_file(tmp_file, table.concat(read_text, '\n'), true)
+  end)
+  after_each(cleanup)
+  it('inserts text from file', function()
+    test_read('read ' .. tmp_file, inserted_middle)
+    eq({ 0, 2, 2, 0 }, fn.getpos('.'))
+  end)
+  it('inserts text from shell', function()
+    test_read('read !cat ' .. tmp_file, inserted_middle)
+    eq({ 0, 3, 3, 0 }, fn.getpos('.'))
+  end)
+  it('inserts text from command', function()
+    local make_lines = string.format('let lines="%s"', table.concat(read_text, '\\n'))
+    execute(make_lines)
+    test_read('read :echo lines', inserted_middle)
```

Comment:
also need a test that checks behavior with `|`
```suggestion
    test_read('read :foo | bar | baz', inserted_middle)
```

Reply:
still need this